spectrum-offchain/src/box_resolver/persistence.rs

```diff
@@ -1,77 +1,58 @@
     async fn get_last_confirmed<'a>(&self, id: TEntity::TEntityId) -> Option<Confirmed<TEntity>>
     where
         <TEntity as OnChainEntity>::TEntityId: 'a;
+    /// Get last unconfirmed state of the given entity.
     async fn get_last_unconfirmed<'a>(&self, id: TEntity::TEntityId) -> Option<Unconfirmed<TEntity>>
     where
         <TEntity as OnChainEntity>::TEntityId: 'a;
+    /// Persist predicted state of the entity.
     async fn put_predicted<'a>(&mut self, entity: Traced<Predicted<TEntity>>)
     where
         Traced<Predicted<TEntity>>: 'a;
+    /// Persist confirmed state of the entity.
     async fn put_confirmed<'a>(&mut self, entity: Confirmed<TEntity>)
     where
         Traced<Predicted<TEntity>>: 'a;
+    /// Persist unconfirmed state of the entity.
     async fn put_unconfirmed<'a>(&mut self, entity: Unconfirmed<TEntity>)
     where
         Traced<Predicted<TEntity>>: 'a;
-    async fn invalidate<'a>(&mut self, eid: TEntity::TEntityId, sid: TEntity::TStateId)
+    /// Invalidate particular state of the entity.
+    async fn invalidate<'a>(&mut self, sid: TEntity::TStateId)
     where
         <TEntity as OnChainEntity>::TStateId: 'a,
         <TEntity as OnChainEntity>::TEntityId: 'a;
+    /// False-positive analog of `exists()`.
+    async fn may_exist<'a>(&self, sid: TEntity::TStateId) -> bool
+    where
+        <TEntity as OnChainEntity>::TStateId: 'a;
     async fn get_state<'a>(&self, sid: TEntity::TStateId) -> Option<TEntity>
     where
         <TEntity as OnChainEntity>::TStateId: 'a;
 }
 
-static PREDICTED_KEY_PREFIX: &str = "predicted:prevState:";
-static LAST_PREDICTED_KEY_PREFIX: &str = "predicted:last:";
-static LAST_CONFIRMED_KEY_PREFIX: &str = "confirmed:last:";
-static LAST_UNCONFIRMED_KEY_PREFIX: &str = "unconfirmed:last:";
-
-pub fn predicted_key_bytes<T: Serialize>(id: &T) -> Vec<u8> {
-    let mut key_bytes = bincode::serialize(PREDICTED_KEY_PREFIX).unwrap();
-    let id_bytes = bincode::serialize(&id).unwrap();
-    key_bytes.extend_from_slice(&id_bytes);
-    key_bytes
-}
-
-pub fn last_predicted_key_bytes<T: Serialize>(id: &T) -> Vec<u8> {
-    let mut key_bytes = bincode::serialize(LAST_PREDICTED_KEY_PREFIX).unwrap();
-    let id_bytes = bincode::serialize(&id).unwrap();
-    key_bytes.extend_from_slice(&id_bytes);
-    key_bytes
-}
-
-pub fn last_confirmed_key_bytes<T: Serialize>(id: &T) -> Vec<u8> {
-    let mut key_bytes = bincode::serialize(LAST_CONFIRMED_KEY_PREFIX).unwrap();
-    let id_bytes = bincode::serialize(&id).unwrap();
-    key_bytes.extend_from_slice(&id_bytes);
-    key_bytes
-}
-
-pub fn last_unconfirmed_key_bytes<T: Serialize>(id: &T) -> Vec<u8> {
-    let mut key_bytes = bincode::serialize(LAST_UNCONFIRMED_KEY_PREFIX).unwrap();
-    let id_bytes = bincode::serialize(&id).unwrap();
-    key_bytes.extend_from_slice(&id_bytes);
-    key_bytes
-}
-
 /// NOTE: do not run the tests in parallel! Rocksdb will complain about lock contention. Use the
```

Comment:
With the randomness used in `rocks_db_client` below, we can remove this comment.